contribs/src/main/java/com/netflix/conductor/contribs/queue/kafka/KafkaObservableQueue.java

```diff
@@ -0,0 +1,519 @@
+		return queueName;
+	}
+
+	/**
+	 * Polls the topics and retrieve the messages for all consumers of the topic.
+	 * 
+	 * @return List of messages
+	 */
+	@VisibleForTesting()
+	List<Message> receiveMessages() {
+		List<Message> messages = new ArrayList<>();
+		/*
+		 * Accumulate messages for all consumers as if there is only one consumer
+		 */
+		if(this.consumers != null) {
+			for (KafkaConsumer<String, String> consumer : this.consumers) {
+				messages.addAll(receiveMessages(consumer));
+			}
+		}
+		return messages;
+	}
+
+	/**
+	 * Polls the topics and retrieve the messages for a consumer.
+	 * 
+	 * @return List of messages
+	 */
+	@VisibleForTesting()
+	List<Message> receiveMessages(KafkaConsumer<String, String> consumer) {
+		List<Message> messages = new ArrayList<>();
+		try {
+
+			ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(pollTimeoutInMs));
+
+			if (records.count() == 0) {
+				return messages;
+			}
+
+			logger.info("polled {} messages from kafka topic.", records.count());
+			records.forEach(record -> {
+				logger.debug("Consumer Record: " + "key: {}, " + "value: {}, " + "partition: {}, " + "offset: {}",
+						record.key(), record.value(), record.partition(), record.offset());
+				String id = record.key() + ":" + record.topic() + ":" + record.partition() + ":" + record.offset();
+				Message message = new Message(id, String.valueOf(record.value()), "");
+				messages.add(message);
+			});
+		} catch (KafkaException e) {
+			logger.error("kafka consumer message polling failed.", e);
+		}
+		return messages;
+	}
+
+	/**
+	 * Publish the messages to the given topic.
+	 * 
+	 * @param messages
+	 */
+	@VisibleForTesting()
+	void publishMessages(List<Message> messages) {
+
+		if (messages == null || messages.isEmpty()) {
+			return;
+		}
+		for (Message message : messages) {
+			final ProducerRecord<String, String> record = new ProducerRecord<>(queueName, message.getId(),
+					message.getPayload());
+
+			RecordMetadata metadata;
+			try {
+				metadata = this.producer.send(record).get();
+				logger.debug("Producer Record: key {}, value {}, partition {}, offset {}", record.key(), record.value(),
+						metadata.partition(), metadata.offset());
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				logger.error("Publish message to kafka topic {} failed with an error: {}", queueName, e.getMessage(),
+						e);
+			} catch (ExecutionException e) {
+				logger.error("Publish message to kafka topic {} failed with an error: {}", queueName, e.getMessage(),
+						e);
+				throw new ApplicationException(ApplicationException.Code.INTERNAL_ERROR, "Failed to publish the event");
+			}
+		}
+		logger.info("Messages published to kafka topic {}. count {}", queueName, messages.size());
+
+	}
+
+
+	@Override
+	public void processFailures(List<Message> messages, EventProcessingFailures failures) {
+		processFailures(messages, failures.getTransientFailures());
```

Comment:
The pattern here seems to be similar to DLQ (Dead Letter Queue). Transient failures are not acked, which means they are intended to be retried. Should those be pushed to the "error queue" as well?

Reply:
@aravindanr The intent is that by not acking the offsets are not bumped up so the consumer will automatically 'retry' because it will get the document at the old current offset (the same document that had the transient failure).